homeassistant/components/harmony/remote.py

```diff
@@ -81,49 +74,73 @@
+            _LOGGER.debug("Discovered host already known: %s", matching_name)
             return
+
+        import_config = {
+            ATTR_DELAY_SECS: matching_configuration_entry.get(
+                ATTR_DELAY_SECS, DEFAULT_DELAY_SECS
+            ),
+            ATTR_ACTIVITY: matching_configuration_entry.get(ATTR_ACTIVITY),
+            CONF_NAME: matching_name,
+            CONF_HOST: discovery_info.get(CONF_HOST),
+        }
+        _LOGGER.info(
+            "The hub with name '%s' has been found to have address '%s'.",
+            import_config[CONF_NAME],
+            import_config[CONF_HOST],
+        )
+        # The name matches one of the names that was provided in the yaml
+        # We fall though and will proceed to import the config entry
     elif CONF_HOST in config:
-        host = (config.get(CONF_NAME), config.get(CONF_HOST), config.get(CONF_PORT))
-        activity = config.get(ATTR_ACTIVITY)
-        delay_secs = config.get(ATTR_DELAY_SECS)
+        # Here we are loading the yaml config and CONF_HOST is defined
+        import_config = config
     else:
+        # Here we are loading the yaml config and CONF_HOST is NOT defined
+        #
+        # Cache the device config entry so discovery can add it to the list
+        # of entries without CONF_HOST it will auto import as soon as this function
+        # is called with discovery_info that contains a matching CONF_NAME
+        _LOGGER.info(
+            "The hub with name '%s' will be imported upon discovery.",
+            config[CONF_NAME],
+        )
         hass.data[CONF_DEVICE_CACHE].append(config)
         return
 
-    name, address, port = host
-    _LOGGER.info(
-        "Loading Harmony Platform: %s at %s:%s, startup activity: %s",
-        name,
-        address,
-        port,
-        activity,
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=import_config
+        )
     )
 
-    harmony_conf_file = hass.config.path(f"harmony_{slugify(name)}.conf")
-    try:
-        device = HarmonyRemote(
-            name, address, port, activity, harmony_conf_file, delay_secs
-        )
-        if not await device.connect():
-            raise PlatformNotReady
 
-        DEVICES.append(device)
-        async_add_entities([device])
-        register_services(hass)
-    except (ValueError, AttributeError):
-        raise PlatformNotReady
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
+):
+    """Set up the Harmony config entry."""
+
+    device = hass.data[DOMAIN][entry.entry_id]
+
+    _LOGGER.info("Harmony Remote: %s", device)
+
+    async_add_entities([device])
+    register_services(hass)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Shutdown a harmony remote for removal."""
+
+    device = hass.data[DOMAIN][entry.entry_id]
+    await device.shutdown()
```

Comment:
This can just as well be done in __init__.py unload_entry

Reply:
done.